addons/delivery/models/stock_picking.py

```diff
@@ -186,12 +186,32 @@ def send_to_shipper(self):
         msg = _("Shipment sent to carrier %s for shipping with tracking number %s<br/>Cost: %.2f %s") % (self.carrier_id.name, self.carrier_tracking_ref, self.carrier_price, order_currency.name)
         self.message_post(body=msg)
 
+    def _get_new_delivery_price(self):
+        if self.carrier_id.integration_level != 'rate_and_ship':
+            res = self.carrier_id.rate_shipment(self.sale_id)
+            if res['success']:
+                self.carrier_price = res['price']
+            else:
+                raise UserError(_("Unable to update the delivery price because of: ") + res['error_message'])
+
     @api.multi
     def _add_delivery_cost_to_so(self):
         self.ensure_one()
         sale_order = self.sale_id
         if sale_order.invoice_shipping_on_delivery:
-            sale_order._create_delivery_line(self.carrier_id, self.carrier_price)
+            self._get_new_delivery_price()
```

Comment:
```
# if there isn't a delivery line on the SO yet
if sale_order.invoice_shipping_on_delivery:
    self._get_new_delivery_price()  # will `self.carrier_price`
    # we force `price_unit_is_estimated` because we want the delivery line to have a price_unit
    sale_order._create_delivery_line(self.carrier_id, self.carrier_price, price_unit_is_estimated=False)
```